README.md
# Bytes Utility Library for Java

Bytes is a utility library that makes it easy to **create**, **parse**, **transform**,
**validate** and **convert** byte arrays in Java. It's main class `Bytes` is
a collections of bytes and the main API. It supports [endianness](https://en.wikipedia.org/wiki/Endianness)
as well as **copy-on-write** and **mutable** access, so the caller may decide to favor
performance. This can be seen as combination of the features provided by
[`BigInteger`](https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html),
[`ByteBuffer`](https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html) but
providing a lot of additional features on the micro and macro level of byte arrays (similar to Okio's [ByteString](https://github.com/square/okio)). The main goal is to minimize the need
to blindly paste code snippets from
[s](https://stackoverflow.com/questions/140131/convert-a-string-representation-of-a-hex-dump-to-a-byte-array-using-java)
[t](https://stackoverflow.com/questions/12893758/how-to-reverse-the-byte-array-in-java)
[a](https://stackoverflow.com/questions/3329163/is-there-an-equivalent-to-memcpy-in-java)
[c](https://stackoverflow.com/questions/5513152/easy-way-to-concatenate-two-byte-arrays)
[k](https://stackoverflow.com/questions/1936857/convert-integer-into-byte-array-java)
[o](https://stackoverflow.com/questions/14243922/java-xor-over-two-arrays)
[v](https://stackoverflow.com/questions/28997781/bit-shift-operations-on-a-byte-array-in-java)
[e](https://stackoverflow.com/questions/13109588/base64-encoding-in-java)
[r](https://stackoverflow.com/questions/2091454/byte-to-inputstream-or-outputstream)
[f](https://stackoverflow.com/questions/3736058/java-object-to-byte-and-byte-to-object-converter-for-tokyo-cabinet)
[l](https://stackoverflow.com/questions/4231674/converting-an-array-of-bytes-to-listbyte)
[o](https://stackoverflow.com/questions/28703273/sorting-byte-arrays-in-numeric-order)
[w](https://stackoverflow.com/questions/4385623/bytes-of-a-string-in-java)
[.](https://stackoverflow.com/questions/23360692/byte-position-in-java)
[c](https://stackoverflow.com/questions/11437203/byte-array-to-int-array)
[o](https://stackoverflow.com/a/9670279/774398)
[m](https://stackoverflow.com/questions/1519736/random-shuffling-of-an-array)

[![Maven Central](https://img.shields.io/maven-central/v/at.favre.lib/bytes)](https://mvnrepository.com/artifact/at.favre.lib/bytes)
[![Github Actions](https://github.com/patrickfav/bytes-java/actions/workflows/build_deploy.yml/badge.svg)](https://github.com/patrickfav/bytes-java/actions)
[![Javadocs](https://www.javadoc.io/badge/at.favre.lib/bytes.svg)](https://www.javadoc.io/doc/at.favre.lib/bytes)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=patrickfav_bytes-java&metric=coverage)](https://sonarcloud.io/summary/new_code?id=patrickfav_bytes-java)
[![Maintainability Rating](https://sonarcloud.io/api/project_badges/measure?project=patrickfav_bytes-java&metric=sqale_rating)](https://sonarcloud.io/summary/new_code?id=patrickfav_bytes-java)
[![Reliability Rating](https://sonarcloud.io/api/project_badges/measure?project=patrickfav_bytes-java&metric=reliability_rating)](https://sonarcloud.io/summary/new_code?id=patrickfav_bytes-java)

Its main features include:

* **Creation** from a wide variety of sources: multiple arrays, integers, [streams](https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html), random, strings, files, uuid, ...
* **Transformation** with many built-in: append, [xor](https://en.wikipedia.org/wiki/Exclusive_or), [and](https://en.wikipedia.org/wiki/Logical_conjunction), [hash](https://en.wikipedia.org/wiki/Cryptographic_hash_function), [shifts](https://en.wikipedia.org/wiki/Bitwise_operation#Bit_shifts), shuffle, reverse, [checksum](https://en.wikipedia.org/wiki/Checksum), ...
* **Validators** with the ability to arbitrarily combine multiple ones with logical expressions
* **Parsing and Encoding** in most common binary-to-text-encodings: [hex](https://en.wikipedia.org/wiki/Hexadecimal), [base32](https://en.wikipedia.org/wiki/Base32), [base64](https://en.wikipedia.org/wiki/Base64), ...
* **Immutable, Mutable and Read-Only** versions
* **Handling Strings** with encoding and normalizing strings for arbitrary charset
* **Utility Features** like `indexOf`, `count`, `isEmpty`, `bitAt`, `contains` ...
* **Flexibility** provide your own Transformers, Validators and Encoders

The code is compiled with target [Java 7](https://en.wikipedia.org/wiki/Java_version_history#Java_SE_7) to keep backwards compatibility with *Android* and older *Java* applications.
It is lightweight as it does not require any additional dependencies.

## Quickstart

Add dependency to your `pom.xml` ([check latest release](https://github.com/patrickfav/bytes-java/releases)):

```xml
<dependency>
    <groupId>at.favre.lib</groupId>
    <artifactId>bytes</artifactId>
    <version>{latest-version}</version>
</dependency>
```

_Note:_ There is a byte-code optimized version (powered by [ProGuard](https://www.guardsquare.com/en/products/proguard)) which can be used with [classifier](https://maven.apache.org/pom.html#Maven_Coordinates) 'optimized'. This may have issues so use at your own risk.

Some simple examples:

```java
Bytes b = Bytes.wrap(someByteArray);  //reuse given reference
b.copy().reverse(); //reverse the bytes on a copied instance
String hex = b.encodeHex(); //encode base16/hex
```

```java
Bytes b = Bytes.parseHex("0ae422f3");  //parse from hex string
int result = b.toInt(); //get as signed int
```

```java
Bytes b = Bytes.from(array1);  //create from copy of array1
b.resize(2).xor(array2); //shrink to 2 bytes and xor with other array
byte[] result = b.array(); //get as byte array
```

## API Description

Per default the instance is **semi-immutable**, which means any transformation will
create a copy of the internal array (it is, however, possible to get and
modify the internal array). There is a **mutable** version which supports
in-place modification for better performance and a **read-only** version which
restricts the access to the internal array.

### Constructors

There are 3 basic constructors:

 * `wrap()` which reuses the given array reference; this is equivalent to `ByteBuffer.wrap()`
 * `from()` which always creates a new internal array reference (i.e. a copy of the passed reference)
 * `parse()` which parses from binary-text-encoded strings (see other section)

Here is a simple example to show the difference:

```java
byte[] myArray = ...
Bytes bWrap = Bytes.wrap(myArray);
assertSame(myArray, bWrap.array());

byte[] myArray2 = ...
Bytes bFrom = Bytes.from(myArray2);
assertNotSame(myArray2, bFrom.array());
assertArrayEquals(myArray2, bFrom.array());
```

The following code is equivalent:

```java
Bytes.wrap(myArray).copy() ~ Bytes.from(myArray)
```

#### More Constructors

For a **null-safe version**, which uses the empty array in case of a null byte array:

```java
Bytes.wrapNullSafe(null);
Bytes.fromNullSafe(null);
```

**Concatenating** of multiple byte arrays or bytes:

```java
Bytes.from(array1, array2, array3);
Bytes.from((byte) 0x01, (byte) 0x02, (byte) 0x03);
```

Creating byte arrays from **primitive integer** types and arrays:

```java
Bytes.from(8);  //00000000 00000000 00000000 00001000
Bytes.from(1897621543227L);
Bytes.from(1634, 88903, 77263);
Bytes.from(0.7336f, -87263.0f);
Bytes.from(0.8160183296, 3984639846.0);
```

Initializing **empty arrays** of arbitrary length:

```java
Bytes.allocate(16);
Bytes.allocate(4, (byte) 1); //fill with 0x01
Bytes.empty(); //creates zero length byte array
```

Creating cryptographically secure **random** byte arrays:

```java
Bytes.random(12);
```

Creating cryptographically unsecure **random** byte arrays for e.g. testing:

```java
Bytes.unsecureRandom(12, 12345L); // using seed makes it deterministic
```

Reading byte content of encoded `String`s:

```java
Bytes.from(utf8String)
Bytes.from(utf8StringToNormalize, Normalizer.Form.NFKD) //normalizes unicode
Bytes.from(asciiString, StandardCharset.US_ASCII) //any charset
```

And other types:

```java
Bytes.from(byteInputStream); //read whole java.io.InputStream
Bytes.from(byteInputStream, 16); //read java.io.InputStream with length limitation
Bytes.from(byteList); //List<Byte> byteList = ...
Bytes.from(myBitSet); //java.util.BitSet myBitSet = ...
Bytes.from(bigInteger); //java.math.BigInteger
Bytes.from(file); //reads bytes from any java.io.File
Bytes.from(dataInput, 16); //reads bytes from any java.io.DataInput
Bytes.from(UUID.randomUUID()); //read 16 bytes from UUID
```

For parsing binary-text-encoded strings, see below.

### Transformers

Transformers transform the internal byte array. It is possible to create
custom transformers if a specific feature is not provided by the default
 implementation (see `BytesTransformer`). Depending on the type (mutable vs
 immutable) and transformer it will overwrite the internal byte array
 or always create a copy first.

```java
Bytes result = Bytes.wrap(array1).transform(myCustomTransformer);
```

#### Built-In Transformers

For **appending** byte arrays or primitive integer types to current instances.
*Note:* this will create a new copy of the internal byte array; for dynamically
 growing byte arrays see `ByteArrayOutputStream`.

```java
Bytes result = Bytes.wrap(array1).append(array2);
Bytes result = Bytes.wrap(array1).append(1341);
Bytes result = Bytes.wrap(array1).append((byte) 3);
Bytes result = Bytes.wrap(array1).append("some string");
```

**Bitwise operations**: XOR, OR, AND, NOT as well as left and right shifts and switching bits:

```java
Bytes.wrap(array).xor(array2); // 0010 0011 xor() 1011 1000 = 1001 1011
Bytes.wrap(array).or(array2); // 0010 0011 or() 1101 0100 = 1111 0111
Bytes.wrap(array).and(array2); // 0010 0011 and() 1011 1000 = 0010 0000
Bytes.wrap(array).not(); // 0010 0011 negate() = 1101 1100
Bytes.wrap(array).leftShift(8);
Bytes.wrap(array).rightShift(8);
Bytes.wrap(array).switchBit(3, true);
```

**Copy** operations, which copies the internal byte array to a new instance:

```java
Bytes copy = Bytes.wrap(array).copy();
Bytes copy = Bytes.wrap(array).copy(3, 17); //copy partial array
```

**Resizing** the internal byte array:

```java
Bytes resized = Bytes.wrap(array).resize(3); //from {3, 9, 2, 1} to {9, 2, 1}
```

**Hashing** the internal byte array using the [`MessageDigest`](https://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html) Java crypto API:

```java
Bytes hash = Bytes.wrap(array).hashSha256();
Bytes hash = Bytes.wrap(array).hashSha1();
Bytes hash = Bytes.wrap(array).hashMd5();
Bytes hash = Bytes.wrap(array).hash("SHA-512");
```

**Reversing** of the byte order in the array

```java
Bytes result = Bytes.wrap(array).reverse();
```

#### Additional Transformers

More transformers can be accessed through the `BytesTransformers`, which
can be statically imported for a less verbose syntax:

```java
import static at.favre.lib.bytes.BytesTransformers.*;
```

**HMAC** used to calculate [keyed-hash message authentication code](https://en.wikipedia.org/wiki/HMAC):

```java
Bytes.wrap(array).transform(hmacSha256(macKey32Byte));
Bytes.wrap(array).transform(hmacSha1(macKey20Byte));
Bytes.wrap(array).transform(hmac(macKey16Byte,"HmacMd5"));
```

**Checksum** can be calculated or automatically appended:

```java
Bytes.wrap(array).transform(checksumAppendCrc32());
Bytes.wrap(array).transform(checksumCrc32());
Bytes.wrap(array).transform(checksum(new Adler32(), ChecksumTransformer.Mode.TRANSFORM, 4));
```

**GZip compression** is supported by [`GZIPInputStream`](https://docs.oracle.com/javase/7/docs/api/java/util/zip/GZIPInputStream.html):

```java
Bytes compressed = Bytes.wrap(array).transform(compressGzip());
Bytes decompressed = compressed.transform(decompressGzip());
```

**Sorting** of individual bytes with either [`Comparator`](https://docs.oracle.com/javase/7/docs/api/java/util/Comparator.html) or natural order:

```java
Bytes.wrap(array).transform(sort()); // 0x00 sorts after 0xff
Bytes.wrap(array).transform(sortUnsigned()); // 0xff sorts after 0x00
Bytes.wrap(array).transform(sort(byteComparator));
```

**Shuffling** of individual bytes:

```java
Bytes.wrap(array).transform(shuffle());
```

### Parser and Encoder for Binary-Text-Encodings

This library can parse and encode a variety of encodings: binary, decimal, [octal](https://en.wikipedia.org/wiki/Octal),
[hex](https://en.wikipedia.org/wiki/Hexadecimal) and
[base64](https://en.wikipedia.org/wiki/Base64). Additionally custom parsers are supported by providing your own
implementation:

```java
Bytes.parse("8sK;S*j=r", base85Decoder);
Bytes.encode(base85Encoder);
 ```

**Hex** can be upper and lowercase and also supports `0x` prefix for parsing:

```java
Bytes.parseHex("a0e13eaa1a")
Bytes.parseHex("0xA0E1")

Bytes.from(array).encodeHex() //a0e13eaa1a
 ```

This lib has it's own build in **Base64** encoder:

```java
Bytes.parseBase64("SpT9/x6v7Q==");

Bytes.from(array).encodeBase64(); //"SpT9/x6v7Q=="
Bytes.from(array).encodeBase64Url(); //"SpT9_x6v7Q=="
 ```

also a **Base32** encoder (using the RFC4648 non-hex alphabet):

```java
Bytes.parseBase32("MZXQ====");
Bytes.from(array).encodeBase32();
 ```

Additionally the following radix encodings are supported:

```java
Bytes.from(array).encodeBinary(); //1110110110101111
Bytes.from(array).encodeDec(); //20992966904426477
Bytes.from(array).encodeOctal(); //1124517677707527755
Bytes.from(array).encodeRadix(36); //5qpdvuwjvu5
```

### Handling Strings

You can easily get the **UTF-8 encoded version** of a string with

```java
String s = "...";
Bytes.from(s);
```

or get the **[normalized version](https://en.wikipedia.org/wiki/Unicode_equivalence)**,
which is the recommended way to convert e.g. user names

```java
String pwd = "ℌH";
Bytes.from(pwd, Normalizer.Form.NFKD); //would be "HH" normalized
```

or get as any other **[character encodings](https://en.wikipedia.org/wiki/Character_encoding)**

```java
String asciiString = "ascii";
Bytes.from(asciiString, StandardCharsets.US_ASCII);
```

To easily append a string to an byte array you can do

```java
String userPwdHash = ...;
Bytes.from(salt).append(userPwd).hashSha256();
```

### Utility Methods

Methods that return additional information about the instance.

Finding occurrence of specific bytes:

```java
Bytes.wrap(array).contains((byte) 0xE1);
Bytes.wrap(array).indexOf((byte) 0xFD);
Bytes.wrap(array).indexOf(new byte[] {(byte) 0xFD, 0x23});
Bytes.wrap(array).indexOf((byte) 0xFD, 5); //search fromIndex 5
Bytes.wrap(array).lastIndexOf((byte) 0xAE);
Bytes.wrap(array).startsWith(new byte[] {(byte) 0xAE, 0x32});
Bytes.wrap(array).endsWidth(new byte[] {(byte) 0xAE, 0x23});
```

Length checks:

```java
Bytes.wrap(array).length();
Bytes.wrap(array).lengthBit(); //8 * array.length
Bytes.wrap(array).isEmpty();
```

Accessing part of the array as primitives from arbitrary position:

```java
Bytes.wrap(array).bitAt(4); // 0010 1000 -> false
Bytes.wrap(array).byteAt(14); // 1111 1111 -> -1
Bytes.wrap(array).unsignedByteAt(14); // 1111 1111 -> 255
Bytes.wrap(array).intAt(4);
Bytes.wrap(array).longAt(6);
```

And others:

```java
Bytes.wrap(array).count(0x01); //occurrences of 0x01
Bytes.wrap(array).count(new byte[] {0x01, 0xEF}); //occurrences of pattern [0x01, 0xEF]
Bytes.wrap(array).entropy();
```

Of course all standard Java Object methods are implemented including:
`hashCode()`, `equals()`, `toString()` as well as it being
[`Comparable`](https://docs.oracle.com/javase/7/docs/api/java/lang/Comparable.html).
In addition there is a constant time `equalsConstantTime()` method, see [here](https://codahale.com/a-lesson-in-timing-attacks/) why this
might be useful.

The `toString()` methods only shows the length and a preview of maximal 8 bytes:

```
16 bytes (0x7ed1fdaa...12af000a)
```

Bytes also implements the `Iterable` interface, so it can be used in a
foreach loop:

```java
for (Byte aByte : bytesInstance) {
    ...
}
```

The `equals` method has overloaded versions for `byte[]`, `Byte[]` and `ByteBuffer` which can be used to directly
compare the inner array:

```java
byte[] primitiveArray1 = ...
byte[] primitiveArray2 = ...
Bytes.wrap(primitiveArray1).equals(primitiveArray2); //compares primitiveArray1 with primitiveArray2
```

### Validation

A simple validation framework which can be used to check the internal byte array:

```java
import static at.favre.lib.bytes.BytesValidators.*;

Bytes.wrap(new byte[]{8, 3, 9}).validate(startsWith((byte) 8), atLeast(3)); // true
```

This is especially convenient when combining validators:

```java
Bytes.wrap(new byte[]{0, 1}).validate(atMost(2), notOnlyOf((byte)  0)); // true
```

Validators also support nestable logical expressions AND, OR as well as NOT:

```java
Bytes.allocate(0).validate(or(exactLength(1), exactLength(0))) //true
Bytes.allocate(19).validate(and(atLeast(3), atMost(20))) //true
Bytes.allocate(2).validate(not(onlyOf((byte) 0))); //false
```

Nesting is also possible:

```java
assertTrue(Bytes.allocate(16).validate(
                or(
                   and(atLeast(8),not(onlyOf(((byte) 0)))),
                   or(exactLength(16), exactLength(12))))); // true
```

### Converting

The internal byte array can be converted or exported into many different formats.
There are 2 different kinds of converters:

* Ones that create a new type which **reuses the same shared memory**
* Ones that create a **copy** of the internal array, which start with `to*`

#### Shared Memory Conversion

Not technically a conversation, but it is of course possible to access the internal array:

```java
Bytes.wrap(array).array();
```

Conversion to [`InputStream`](https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html)
 and [`ByteBuffer`](https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html):

```java
Bytes.wrap(array).inputStream();
Bytes.wrap(array).buffer();
```

If you just want a duplicated instance, sharing the same array:

```java
Bytes.wrap(array).duplicate();
```

For the conversion to read-only and mutability, see below.

#### Copy Conversion

To primitives (if the internal array is not too long)

```java
Bytes.wrap(array).toByte();
Bytes.wrap(array).toUnsignedByte();
Bytes.wrap(array).toInt();
Bytes.wrap(array).toDouble();
```

To primitive arrays

```java
Bytes.wrap(array).toIntArray(); // of type int[]
Bytes.wrap(array).toLongArray(); // of type long[]
```

To other collections

```java
Bytes.wrap(array).toList(); // of type List<Byte>
Bytes.wrap(array).toBoxedArray(); // of type Byte[]
Bytes.wrap(array).toBitSet(); //of type java.util.BitSet
```

to [`BigInteger`](https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html) of course

```java
Bytes.wrap(array).toBigInteger();
```

and others

```java
Bytes.wrap(array).toUUID(); // convert 16 byte to UUID
Bytes.wrap(array).toCharArray(StandardCharsets.UTF-8); // converts to encoded char array
```

### Mutable and Read-Only

Per default the instance is immutable, i.e. every transformation will create a
a new internal byte array (very similar to the API of `BigInteger`). While
this is usually the default way to design such a construct because it shows
[various advantages](https://softwareengineering.stackexchange.com/questions/151733/if-immutable-objects-are-good-why-do-people-keep-creating-mutable-objects)
this can introduce a major performance issue when handling big arrays
or many transformations.

#### Mutable Bytes

All transformers (if possible) reuse or overwrite the same internal memory
to avoid unneeded array creation to minimize time and space complexity.
To create a mutable instance just do:

```java
MutableBytes b = Bytes.from(array).mutable();
```

Mutable classes also enable further APIs for directly modify the internal array:

```java
b.setByteAt(3, (byte) 0xF1)
b.overwrite(anotherArray) //directly overwrite given array
b.fill(0x03) // fills with e.g. 3
b.wipe() //fills with zeros
b.secureWipe() //fills with random data
```

Create a immutable version again with:

```java
Bytes b2 = b.immutable();
```

*Note:* a copy will inherit mutability/read-only properties:

```java
Bytes b = Bytes.from(array).mutable().copy();
assertTrue(b.isMutable());
```

##### AutoClosable for try-with-resources

In security-relevant environments it is best practice to wipe the memory of secret data, such as
secret keys. This can be used with Java 7 feature try-with-resource like this:

```java
try (MutableBytes b = Bytes.wrap(aesBytes).mutable()) {
    SecretKey s = new SecretKeySpec(b.array(), "AES");
    ...
}
```

#### Readonly Bytes

On the other hand, if you want a export a instance with limited access,
especially no easy way to alter the internal byte array, read-only instances
may be created by:

```java
Bytes b = Bytes.from(array).readOnly();
```

Every call to the following conversation methods will throw a `ReadOnlyBufferException`:

```java
readOnlyBytes.array();
readOnlyBytes.byteBuffer();
readOnlyBytes.inputStream();
```

## Download

The artifacts are deployed to [Maven Central](https://search.maven.org/).

### Maven

Add the dependency of the [latest version](https://github.com/patrickfav/bytes-java/releases) to your `pom.xml`:

```xml
<dependency>
    <groupId>at.favre.lib</groupId>
    <artifactId>bytes</artifactId>
    <version>{latest-version}</version>
</dependency>
```

### Gradle

Add to your `build.gradle` module dependencies:

    implementation group: 'at.favre.lib', name: 'bytes', version: '{latest-version}'

### Local Jar Library

[Grab jar from latest release.](https://github.com/patrickfav/bytes-java/releases/latest)

### OSGi

The library should be prepared to be used with the OSGi framework with the help of the [bundle plugin](http://felix.apache.org/documentation/subprojects/apache-felix-maven-bundle-plugin-bnd.html).


## Digital Signatures

### Signed Jar

The provided JARs in the Github release page are signed with my private key:


CHANGELOG
# Releases

## 1.6.2

* remove hashCode caching since it could introduce very subtle bugs

## v1.6.1

* now build by JDK 11 and removed errorprone compiler #52
* introduce sonarqube and remove codecov
* improve javadoc by fixing many typos #53
* some small bugfixes

## v1.6.0

* migrate to github actions, codecov and maven central #49
* add `indexOf` (thx @hlyakhovich) #48
* add `toShortArray` (thx @hlyakhovich) #44
* add `from()` constructor from `short` vararg or array (thx @hlyakhovich) #45
* add an automatic module name to support the JPMS (thx @airsquared) #47
* fix warning of junit 4.13 CVE-2020-15250

## v1.5.0

* fix `leftShift()` and `rightShift()` to respect byte order (thx @gfpeltier)
* fix `bitAt()` to respect byte order (thx @gfpeltier)

## v1.4.0

* add `from()` constructor from `float[]`
* add `from()` constructor from `double[]`
* fix throwing `IllegalArgumentException` instead of `IllegalStateException` in `.toUUID()`

## v1.3.0
* improve hex encoder performance by factor 5

## v1.2.0

* let hex decoder accept odd length string #37

## v1.1.0

* add `unsecureRandom()` constructor which creates random for e.g. tests or deterministic randoms
* adds overwrite method to Bytes (thx @JadePaukkunen)
* make project OSGi compatible #36
* add `toFloatArray()` converter #30
* add `toDoubleArray()` converter #30

## v1.0.0

* add `append()` method supporting multiple byte arrays #26
* add `toCharArray()` method which decodes internal byte array to char[] #27
* add `encodeBase64()` supporting padding-less encoding
* add `toIntArray()` converter #28
* add `toLongArray()` converter #29
* add `AutoCloseable` to MutableBytes interface #31
* add `allocate()` as mutable byte static constructor (thx @petrukhnov)

### Breaking

* removed deprecated `toObjectArray()`; use `toBoxedArray()` instead

## v0.8.0

* add radix encoding/parsing and fix radix tests #6, #20
* add support for Base32 RFC4648 non-hex alphabet encoding/parsing #21
* add constructor for `IntBuffer` and `CharBuffer`
* `parse()` methods now expect more flexible `CharSequence` instead of `String` #23
* `from()` constructor reading from `char[]` has new version that accepts offset and length #24
* add `from()` constructor reading file with offset and length parameter #25

### Breaking

* interface `BinaryToTextEncoding.decode()` changed param to `CharSequence` from `String` #23

### Deprecations (will be removed in v1.0+)

* `parseBase36()`/`encodeBase36()` - use `parseRadix(36)`/`encodeRadix(36)` instead

## v0.7.1

* sign AFTER ProGuard so optimized version has correct jar signature

## v0.7.0

 * add `count` method for counting byte arrays (like pattern matching)
 * add dedicated `md5` and `sha1` transformer methods
 * add `from(Inputstream stream, int maxlength)` limiting stream reading constructor #13
 * add indexOf() with `fromIndex` parameter #14
 * add support for base64 url safe encoding #15
 * use EMPTY constant instance for empty byte array to safe memory #16
 * add `startsWith()` and `endsWidth()` methods #12
 * add cache for calculating the hashCode
 * add HMAC byte transformer #11
 * add unsigned sort transformer #17
 * add `.immutable()` converter in MutableBytes #18

## v0.6.0

 * add `encodeCharsetToBytes()` feature #7
 * add new `from(char[] charArray, Charset charset)` constructor with improved logic #8
 * add constructor/converter from/to UUID #9
 * add `empty()` constructor, creating empty byte array

### Deprecations (will be removed in v1.0+)

* `toObjectArray()` renamed to `toBoxedArray()`

## v0.5.0

 * better resource handling for compression
 * add nullSafe from() constructor
 * rename `toObjectArray()` to `toBoxedArray()` (will be removed in 1.0)
 * add appendNullSafe and append string with encoding
 * add proguard optimized version (can be used with classifier 'optimized')
 * add constant time equals
 * fix or() operator using and() internally #2

## v0.4.6

* add appending with strings
* add boolean, float, double and char array constructor
* try to fix radix encoder missing prefixing 0 byte

## v0.4.5

* add nullSafe wrapper
* fix accepting illegal hex string when parsing

## v0.4.4

* add feature for gathering parts of the array as primitives (e.g. intAt(int position))
* add to unsigned byte conversations
* add overloaded equals

## v0.4.3

* add toFloat/toDouble
* add resize mode (from 0 or length)
* fix bitAt bug
* add in place byte array shift feature

## v0.4.2

* add check method if transformer supports inplace
* add contains method
* add Iterable interface
* add setByteAt
* add construction from DataInput

## v0.4.1

* add bitAt() utility
* add hash feature
* add checksum transformer
* add gzip transformer

## v0.4.0

* logical expressions in validators
* more validators

## v0.3.1

* better validation and more creation possibilities
* create from bigInteger, int[], long[] and Byte[]

## v0.3.0

* better type hierarchy with read-only and mutable types
* add validator feature
* add bit switch feature

## v0.2.0

initial version


src/main/java/at/favre/lib/bytes/Base64.java
/*
 * Copyright 2017 Patrick Favre-Bulle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package at.favre.lib.bytes;

import java.util.Arrays;

/**
 * From https://github.com/square/okio/blob/okio-parent-1.15.0/okio/src/main/java/okio/Base64.java
 * under Apache 2 license.
 *
 * @author Alexander Y. Kleymenov
 */
final class Base64 {
    private Base64() {
    }

    private static final byte[] MAP = new byte[]{
            'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
            'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l',
            'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3', '4',
            '5', '6', '7', '8', '9', '+', '/'
    };

    private static final byte[] URL_MAP = new byte[]{
            'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
            'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l',
            'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3', '4',
            '5', '6', '7', '8', '9', '-', '_'
    };

    static byte[] decode(CharSequence in) {
        // Ignore trailing '=' padding and whitespace from the input.
        int limit = in.length();
        for (; limit > 0; limit--) {
            char c = in.charAt(limit - 1);
            if (c != '=' && c != '\n' && c != '\r' && c != ' ' && c != '\t') {
                break;
            }
        }

        // If the input includes whitespace, this output array will be longer than necessary.
        byte[] out = new byte[(int) (limit * 6L / 8L)];
        int outCount = 0;
        int inCount = 0;

        int word = 0;
        for (int pos = 0; pos < limit; pos++) {
            char c = in.charAt(pos);

            int bits;
            if (c >= 'A' && c <= 'Z') {
                // char ASCII value
                //  A    65    0
                //  Z    90    25 (ASCII - 65)
                bits = c - 65;
            } else if (c >= 'a' && c <= 'z') {
                // char ASCII value
                //  a    97    26
                //  z    122   51 (ASCII - 71)
                bits = c - 71;
            } else if (c >= '0' && c <= '9') {
                // char ASCII value
                //  0    48    52
                //  9    57    61 (ASCII + 4)
                bits = c + 4;
            } else if (c == '+' || c == '-') {
                bits = 62;
            } else if (c == '/' || c == '_') {
                bits = 63;
            } else if (c == '\n' || c == '\r' || c == ' ' || c == '\t') {
                continue;
            } else {
                throw new IllegalArgumentException("invalid character to decode: " + c);
            }

            // Append this char's 6 bits to the word.
            word = (word << 6) | (byte) bits & 0xff;

            // For every 4 chars of input, we accumulate 24 bits of output. Emit 3 bytes.
            inCount++;
            if (inCount % 4 == 0) {
                out[outCount++] = (byte) (word >> 16);
                out[outCount++] = (byte) (word >> 8);
                out[outCount++] = (byte) word;
            }
        }

        int lastWordChars = inCount % 4;
        if (lastWordChars == 1) {
            // We read 1 char followed by "===". But 6 bits is a truncated byte! Fail.
            return null;
        } else if (lastWordChars == 2) {
            // We read 2 chars followed by "==". Emit 1 byte with 8 of those 12 bits.
            word = word << 12;
            out[outCount++] = (byte) (word >> 16);
        } else if (lastWordChars == 3) {
            // We read 3 chars, followed by "=". Emit 2 bytes for 16 of those 18 bits.
            word = word << 6;
            out[outCount++] = (byte) (word >> 16);
            out[outCount++] = (byte) (word >> 8);
        }

        // If we sized our out array perfectly, we're done.
        if (outCount == out.length) return out;

        // Copy the decoded bytes to a new, right-sized array.
        return Arrays.copyOfRange(out, 0, outCount);
    }

    static byte[] encode(byte[] in) {
        return encode(in, false, true);
    }

    static byte[] encode(byte[] in, boolean urlSafe, boolean usePadding) {
        return encode(in, urlSafe ? URL_MAP : MAP, usePadding);
    }

    private static byte[] encode(byte[] in, byte[] map, boolean usePadding) {
        int length = outLength(in.length, usePadding);
        byte[] out = new byte[length];
        int index = 0, end = in.length - in.length % 3;
        for (int i = 0; i < end; i += 3) {
            out[index++] = map[(in[i] & 0xff) >> 2];
            out[index++] = map[((in[i] & 0x03) << 4) | ((in[i + 1] & 0xff) >> 4)];
            out[index++] = map[((in[i + 1] & 0x0f) << 2) | ((in[i + 2] & 0xff) >> 6)];
            out[index++] = map[(in[i + 2] & 0x3f)];
        }
        switch (in.length % 3) {
            case 1:
                out[index++] = map[(in[end] & 0xff) >> 2];
                out[index++] = map[(in[end] & 0x03) << 4];
                if (usePadding) {
                    out[index++] = '=';
                    out[index] = '=';
                }
                break;
            case 2:
                out[index++] = map[(in[end] & 0xff) >> 2];
                out[index++] = map[((in[end] & 0x03) << 4) | ((in[end + 1] & 0xff) >> 4)];
                out[index++] = map[((in[end + 1] & 0x0f) << 2)];

                if (usePadding) {
                    out[index] = '=';
                }
                break;
        }
        return out;
    }

    private static int outLength(int srclen, boolean doPadding) {
        int len;
        if (doPadding) {
            len = 4 * ((srclen + 2) / 3);
        } else {
            int n = srclen % 3;
            len = 4 * (srclen / 3) + (n == 0 ? 0 : n + 1);
        }
        return len;
    }
}
